lib/audit/auditor_test.go

```diff
@@ -352,6 +352,309 @@ func TestAudit(t *testing.T) {
 }`,
 	}
 
+	// Regression test for the case where we are promoting
+	// bcdd5657b1edc1a2eb27356f33dd66b9400d4a084209c33461c7a7da0a32ebb3 (etcd
+	// tag "3.4.7-2"). It is crucial that we read fom *ALL* source registries,
+	// not just the first one that has a match for the destination registry
+	// "us.gcr.io/k8s-artifacts-prod" (because both manifests below promote to
+	// that same destination, but only one of them have the child images of
+	// 3.4.7-2).
+	manifests2 := []reg.Manifest{
+		{
+			Registries: []reg.RegistryContext{
+				{
+					Name: "gcr.io/google-containers",
+					Src:  true,
+				},
+				{
+					// Root promotion
+					Name:           "us.gcr.io/k8s-artifacts-prod",
+					ServiceAccount: "foobar@google-containers.iam.gserviceaccount.com",
+				},
+			},
+
+			Images: []reg.Image{
+				{
+					ImageName: "etcd",
+					Dmap: reg.DigestTags{
+						"sha256:12f377200949c25fde1e54bba639d34d119edd7cfcfb1d117526dba677c03c85": {"3.4.7", "3.4.7-0"},
+					},
+				},
+			},
+		},
+		{
+			Registries: []reg.RegistryContext{
+				{
+					Name: "gcr.io/k8s-staging-etcd",
+					Src:  true,
+				},
+				{
+					// Root promotion
+					Name:           "us.gcr.io/k8s-artifacts-prod",
+					ServiceAccount: "foobar@google-containers.iam.gserviceaccount.com",
+				},
+				{
+					// Non-root promotion
+					Name:           "us.gcr.io/k8s-artifacts-prod/kubernetes",
+					ServiceAccount: "foobar@google-containers.iam.gserviceaccount.com",
+				},
+			},
+
+			Images: []reg.Image{
+				{
+					ImageName: "etcd",
+					Dmap: reg.DigestTags{
+						"sha256:bcdd5657b1edc1a2eb27356f33dd66b9400d4a084209c33461c7a7da0a32ebb3": {"3.4.7-2"},
+					},
+				},
+			},
+		},
+	}
+
+	readRepo2 := map[string]string{
```

Comment:
Is there any of this that can be elided, just to make it more readable?  It may be not..

Reply:
I could add some helpers to make test data like this easier to write (and more succinct), but I'm uncomfortable adding another layer of abstraction. Something to think about though, for sure.